Impacts refactor

benmsanderson opened the pull request:
# Monthly Climate Variability, Impact Assessment Framework, and Code Restructure

## Summary

three major enhancements:

1. **Monthly Climate Variability Model**:  stochastic modeling for  climate variability
2. **Climate Impact Assessment Framework**: Extensible system for calculating climate impacts
3. **Code Restructure & Documentation**: Improved organization,  testing, and updated documentation

## New Features

### Monthly Climate Variability (`MeteorNoiseGenerator`)
- PCA/VARX-based modeling preserving spatial-temporal covariance from CMIP6 training data
- Temperature-dependent seasonality that evolves with global warming
- Ensemble generation for uncertainty quantification
- Extend METEOR object to monthly/seasonal with `to_monthly()` method and `train_noise_model()`

### Climate Impact Assessment Framework (`meteor.impacts`)
- Extensible `ImpactCalculator` class
- Example `DegreeDaysCalculator` for heating/cooling degree days
- `ImpactEnsemble` class for statistical analysis across multiple realizations
- General workflows: `create_impact_ensemble()` and `apply_impact_calculator()`

### Noise model
- Integrated noise model training with caching support
- Multiple experiment support for complex scenarios
- Performance optimization with efficient data processing

## Key Files Added/Modified

### New Modules
- `src/meteor/noise_generator.py` - Monthly climate variability modeling
- `src/meteor/impacts/` - impact assessment framework
- `tests/integration/impacts/` - Impact assessment tests
- `create_small_test_data.py` - Low-res test data generation

### Enhanced Modules  
- `src/meteor/meteor.py` - Added `to_monthly()` method
- `src/meteor/cmip6_meteor_data_getter.py` - Added `train_noise_model()`
- Test infrastructure with coarsened climate data to speedup (datagetter is still the killer, though)


## Usage Examples
```

### Basic Monthly Climate Generation
```python
# Train noise model
noise_model = data_getter.train_noise_model(
    experiments=["historical", "ssp245"],
    model="CanESM5",
    variable_name="tas",
    n_modes=8
)

# Generate monthly climate with variability
annual_pred = pattern_model.predict_from_combined_experiment(emissions, concentrations, ["tas"])["tas"]
monthly_base = pattern_model.to_monthly(annual_pred, start_year=1850)
noise = noise_model.generate_realization(global_temp_trajectory, noise_only=True)
monthly_climate = monthly_base + noise

from meteor.impacts import DegreeDaysCalculator, create_impact_ensemble

# Calculate cooling degree days
cdd_calc = DegreeDaysCalculator(base_temperature=18.0, mode="cooling")
ensemble = create_impact_ensemble(
    monthly_base, cdd_calc, noise_model, global_temp, n_realizations=10
)
print(f"Mean CDD: {ensemble.mean():.1f}, Spread: {ensemble.std():.1f}")
```

benmsanderson committed:
first

benmsanderson committed:
first go at integration

benmsanderson committed:
cleaning up

benmsanderson committed:
more cleaning up

benmsanderson committed:
and again.

benmsanderson committed:
move noise model training to noise_generator

benmsanderson committed:
added noise_only functionality and annual-monthly method

benmsanderson committed:
restored temperature effect on harmonics in noise_only mode

benmsanderson committed:
update readme

benmsanderson committed:
Update MONTHLY_NOISE_README.md

benmsanderson committed:
cleanup

benmsanderson committed:
externally specified temperatures

benmsanderson committed:
bug fixes, example notebook with different configs

benmsanderson committed:
notebook bug fixes

maritsandstad committed:
Merging in from base

maritsandstad committed:
Fixing package setup

maritsandstad committed:
Merge pull request #43 from benmsanderson/meteor_noise-merge_base

Meteor noise merge base

benmsanderson committed:
notebook implementation

benmsanderson committed:
tweaks

benmsanderson committed:
tweaks

benmsanderson committed:
tweaks

benmsanderson committed:
tweaks

benmsanderson committed:
tweaks

benmsanderson committed:
first go at refactor

benmsanderson committed:
linting

benmsanderson committed:
linting

benmsanderson committed:
more linting

benmsanderson committed:
hmmm

benmsanderson committed:
tweak

benmsanderson committed:
more tweaks

benmsanderson committed:
gah

benmsanderson committed:
notebook tweaks

benmsanderson committed:
skipping for now

benmsanderson committed:
added native caching to datagetter

benmsanderson committed:
more tests

benmsanderson committed:
and another

benmsanderson committed:
update data getter test

benmsanderson committed:
linting

benmsanderson committed:
fixed paths in tests

benmsanderson committed:
grrr notebook tests...

benmsanderson committed:
Can we just not support 3.9?

benmsanderson committed:
contextlib thing

benmsanderson committed:
last time?

maritsandstad commented on MONTHLY_NOISE_README.md:
```diff
@@ -0,0 +1,473 @@
+
+#### 1. Annual Pattern Scaling
+- **Training**: Learn spatial climate response patterns from CMIP6 experiments (piControl, abrupt-4xCO2, historical, SSPs)
+- **Forcing**: Use simple climate model (CICERO-SCM) to convert emissions to temperature trajectories
```
Should be forcing not temperature for the Forcing point

benmsanderson replied:
killing these files for now

maritsandstad commented on MONTHLY_NOISE_README.md:
```diff
@@ -0,0 +1,473 @@
+from meteor import Cmip6MeteorDataGetter
+
+# Define experiments and variables
+experiments = ["piControl", "abrupt-4xCO2", "historical", "ssp245", "ssp585"]
```
What is ssp585 doing here? We don't need it, or do we plan to predict on that?

maritsandstad commented on MONTHLY_NOISE_README.md:
````diff
@@ -0,0 +1,473 @@
+#### 4.3 Generate Monthly Noise
+```python
+# Create global temperature trajectory for noise generation
+global_temp_trajectory = annual_predictions["ssp245"]["tas"].mean(dim=["lat", "lon"]).values
````
Shouldn't this be a weighted mean? We have functionality for that in prpatt...

benmsanderson replied:
this file is gone now

maritsandstad commented on QUICK_START.md:
```diff
@@ -0,0 +1,117 @@
+# Initialize CMIP6 data access
+data_getter = Cmip6MeteorDataGetter(
+    exps=["piControl", "abrupt-4xCO2", "historical", "ssp245"],
+    flds=["tas", "pr"]
```
Would add the dbe = ["CMIP", "CMIP", "CMIP", "ScenarioMIP"] piece here as that is not exactly intuitive... Maybe something to consider putting under the hood in the future...

maritsandstad commented on tests/test-data/create_small_test_data.py:
Is this file meant to live in the root folder?

benmsanderson replied:
moved

maritsandstad commented on notebooks/Climate_Bench_METEOR.py:
Why was this deleted? Might be semi-useful to keep around...?

benmsanderson replied:
re-added - this was a mistake, tried jupytext syncing and somehow ended up deleting the original

maritsandstad commented on notebooks/noise_cache/CanESM5_tas_noise_model.pkl:
Is this one meant to be shipped, or just git hups-commit remnant?

maritsandstad commented on scripts/demo_noise_only.py:
This script is just a long list of,  albeit somewhat topical, print statements. I have a hard time seeing the point of having it here? numpy and xarray do not need to be imported for this very elaborate version of "Hello world"...

benmsanderson replied:
gone

maritsandstad commented on src/meteor/cmip6_meteor_data_getter.py:
```diff
@@ -274,6 +284,15 @@ def __init__(
+        # Set up caching
+        self.enable_cache = enable_cache
+        if cache_dir is None:
+            cache_dir = os.path.expanduser("~/.meteor/cmip6_cache")
```
There should probably be some clear info somewhere (at least in the README and probably also in logger info that this is being set up here...)

benmsanderson replied:
yes, agree - done both

maritsandstad commented on src/meteor/cmip6_meteor_data_getter.py:
```diff
@@ -274,6 +284,15 @@ def __init__(
+        self.cache_dir = cache_dir
+
+        if self.enable_cache:
+            os.makedirs(self.cache_dir, exist_ok=True)
```
Some error handling here might be warranted. User could send a path to which they don't have write permissions f.i....

benmsanderson replied:
yes, good point

maritsandstad commented on src/meteor/cmip6_meteor_data_getter.py:
```diff
@@ -292,6 +311,175 @@ def __init__(
+
+            data_to_save.to_netcdf(cache_path)
+        except (OSError, ValueError):
+            # Failed to cache, but don't raise error
```
Maybe add a logger warning, though...

benmsanderson replied:
yup

maritsandstad commented on src/meteor/cmip6_meteor_data_getter.py:
```diff
@@ -425,6 +626,17 @@ def get_single_var_mod_data_yearmean(self, exp, fld, model):
         xr.DataArrray
             Data converted from monthly to yearly mean data and including and extra flat ens dimension
         """
+        # Generate cache key
```
Shouldn't this bit depend on the value of self.enable_cache?

benmsanderson replied:
yes, good point

maritsandstad commented on scripts/demo_unified_api.py:
```diff
@@ -0,0 +1,119 @@
+
+# All these models are identical MeteorNoiseGenerator instances
+print(f"Class method result:     {type(noise_model)}")
+print(f"Batch method result:     {type(noise_models['CanESM5']['tas'])}")
```
This line crashes, noise_models seems to be an empty dict...

benmsanderson replied:
file removed

benmsanderson committed:
cleaning up

maritsandstad commented on src/meteor/impacts/impacts_core.py:
Not a huge fan of the very generic name for this file, could wish for something slightly more descriptive. No big deal, though...

benmsanderson replied:
ok - renamed to impacts_core.py

maritsandstad commented on src/meteor/impacts/impacts_core.py:
I think the docstring style here is not completely in line with what we do elsewhere in the repo. Args -> Parameters. I think the style we normally used is called numpy style. copilot is pretty good with that keyword in my experience...

benmsanderson replied:
ok - revising throughout

maritsandstad commented on src/meteor/impacts/calculators/degree_days.py:
```diff
@@ -0,0 +1,296 @@
+                    "Please verify the data is correct."
+                )
+        elif temp_max > 100:
+            # Mixed range - could be Celsius with hot values or unusual units
```
Maybe say that Celsius or Kelvin is expected... Could be Farenheit?

benmsanderson replied:
yup

maritsandstad commented on src/meteor/impacts/ensemble.py:
```diff
@@ -0,0 +1,183 @@
+            stats_data["ensemble_min"] = ensemble_array.min(dim="ensemble_member")
+        elif stat == "max":
+            stats_data["ensemble_max"] = ensemble_array.max(dim="ensemble_member")
+        elif stat.startswith("quantile_"):
```
Maybe call it percentile since that is what we mean?

benmsanderson replied:
updated

maritsandstad commented on src/meteor/impacts/utils.py:
```diff
@@ -0,0 +1,312 @@
+    return actual_lat, actual_lon
+
+
+def calculate_global_mean(
```
Don't we have a global_mean implementation in prpatt. Maybe it shouldn't necessarily be there, but also we probably shouldn't have many global_mean calculation methods floating around in the same library. I think this one is better as weights can also be provided, though, but I think it should live further up, and be shared for the repository.

benmsanderson replied:
removing this, and calling the prpatt version for now

maritsandstad commented on src/meteor/cmip6_meteor_data_getter.py:
```diff
@@ -498,52 +787,211 @@ def make_meteor_training_data_composite(
+        return result
+
+    # pylint: disable=too-many-arguments,too-many-positional-arguments
+    def train_noise_model(
```
I'm still not entirely convinced that having these noise training methods in here makes a lot of sense...

benmsanderson replied:
gone...

maritsandstad commented on src/meteor/meteor.py:
```diff
@@ -456,3 +533,287 @@ def _predict_combined_experiment_from_forcer_series(
                     else:
                         predicted[fld] = predicted[fld] + tmp
         return predicted
+
```
Do we want to keep all these commented out lines around?

benmsanderson replied:
not anymore

maritsandstad commented on src/meteor/noise_generator.py:
```diff
@@ -0,0 +1,750 @@
+        print(f"Model loaded from {filepath}")
+
+    # pylint: disable=missing-type-doc,too-many-arguments,too-many-positional-arguments,too-many-locals
+    @classmethod
```
These class methods are puzzling to me. I fail to see why they can't just live outside of the class and make a noise generator instance etc? It looks fancy, and maybe I am dumb, but if it doesn't have to be this convoluted it probably shouldn't...

benmsanderson replied:
agree...  more complex than is needed.  converted these to simple functions

maritsandstad commented on tests/unit/impacts/test_base.py:
```diff
@@ -0,0 +1,308 @@
+from meteor.impacts.base import ImpactCalculator, ImpactEnsemble, ImpactResult
+
+
+class TestImpactResult:
```
This looks even less like it needs to be a class than what was in test_workflow.py

benmsanderson replied:
removed all but one

maritsandstad commented on tests/integration/impacts/test_workflow.py:
```diff
@@ -0,0 +1,291 @@
+from meteor.impacts.ensemble import apply_impact_calculator, ensemble_statistics
+
+
+class TestMeteorImpactsWorkflow:
```
Why is this a class? It feels a bit extra... I generally ask copilot to make the tests a little less complicated if I can. Maybe

benmsanderson replied:
This one makes sense to keep like this, I think - from an efficiency perspective.  The class structure allows a single setup and shared data between tests.

maritsandstad commented on tests/unit/impacts/test_degree_days.py:
```diff
@@ -0,0 +1,355 @@
+            coords={"month": np.arange(24), "lat": [50, 60], "lon": [0, 10]},
+        )
+
+    def test_init(self):
```
Doesn't use self. Shouldn't be in class

benmsanderson replied:
gone

maritsandstad commented on tests/unit/impacts/test_degree_days.py:
```diff
@@ -0,0 +1,355 @@
+from meteor.impacts.calculators.degree_days import DegreeDaysCalculator
+
+
+class TestDegreeDaysCalculator:
```
I can kind of see the usecase for the class here, but several methods don't use any self for anything and should be pulled out of the class.

maritsandstad commented on tests/unit/impacts/test_utils.py:
Do any of the classes here have a reason to be classes, or is it just some narrative device to separate different thematic parts of the testing? I haven't looked super-closely, but I suspect many if not all of them are just there for no apparent reason...

benmsanderson replied:
ok - removed most of these

maritsandstad commented on tests/unit/test_cmip6_meteor_data_getter.py:
Not a huge fan of imports inside methods... Several imports are repeated between several of the tests. Some of the tests here are also a bit strange and it is a bit difficult to see what the point of them is...

benmsanderson replied:
restructuring to move imports to top.  removing a bunch of tests

maritsandstad commented on tests/unit/test_cmip6_meteor_data_getter.py:
```diff
@@ -48,7 +89,330 @@ def test_get_unique_models():
+    assert hasattr(default_getter, "check_if_model_has_data")
+
+    # Test custom initialization (just verify it doesn't crash)
+    try:
```
Tests inside try, and more or less blanket excepts are not super useful...

benmsanderson replied:
removing try here

maritsandstad commented on tests/unit/test_cmip6_meteor_data_getter.py:
```diff
@@ -65,3 +429,1658 @@ def test_make_meteor_training_data_composite_more_than_one():
+        assert data_getter.flds == []
+        assert data_getter.exps == ["piControl"]
+    except Exception:
+        # If empty fields cause issues, that's expected
```
Then why do we have a test, if failure is perfectly expected and fine?

benmsanderson replied:
removed test

maritsandstad commented on tests/unit/test_cmip6_meteor_data_getter.py:
```diff
@@ -65,3 +429,1658 @@ def test_make_meteor_training_data_composite_more_than_one():
+            pass
+
+
+def test_error_handling_with_oserror():
```
Why? Why are we here? Why are we testing this? And if we are, why are we not really testing much as we allow exceptions, in one case any exception to pass?

benmsanderson replied:
gone, yeah

maritsandstad commented on tests/unit/test_cmip6_meteor_data_getter.py:
```diff
@@ -65,3 +429,1658 @@ def test_make_meteor_training_data_composite_more_than_one():
+        assert len(df) == 0  # Should be empty DataFrames
+
+
+def test_conditional_logic_patterns():
```
Does this test have anything todo with actually testing the code of the data getter? As far as I can see this test is testing if basic python logic works as expected... Why does this repository need to test that?

benmsanderson replied:
gone

maritsandstad reviewed (changes requested):
Go over and tone down weird tests a bit...

benmsanderson committed:
strip out nonsense tests, pull noise model out of datagetter

benmsanderson committed:
more test removals

benmsanderson committed:
re-added climatebench (jupytest test mistake)

benmsanderson committed:
addressing comments, docstrings, error handling, removing redundant class structures

benmsanderson committed:
modest addl tests

benmsanderson committed:
formatting

benmsanderson committed:
Fix import sorting with isort

benmsanderson committed:
linting

benmsanderson committed:
notebook tests

benmsanderson committed:
couple more tests

benmsanderson committed:
linting

benmsanderson committed:
more brutal test purge

benmsanderson committed:
dropping test coverage requirements to 85pct for now

benmsanderson commented:
OK...  since the review:

- removed 40-odd weird tests, dropped required coverage to 85pct.
- removed the superfluous extra readmes
- moved small data script to test data
- restored climatebench
- added gitignore for noise pickles
- renamed impacts core file
- removed class methods from noise-generator
- standardised docstrings to numpy style
- added celcius clarifier to CDD output
- removed global mean calc from impacts model, point to prpatt
- Converted unnecessary test classes to simple functions in test_base.py, test_degree_days.py, test_utils.py
- Moved all imports to top of test files, removed duplicates throughout test suite.
- Added proper error handling for cache directory creation.
- Added comprehensive logging about cache setup in cmip6_meteor_data_getter.py
- Removed demo scripts: demo_noise_only.py, demo_unified_api.py, demo_impacts.py.

maritsandstad commented:
@benmsanderson do you want me to do another review later today (if I can find the time)? BTW, you also need to update the Changelog before we can merge in

benmsanderson commented:
*want* is a strong word

maritsandstad commented:
You _don't_ want me to review the code?

benmsanderson commented:
I do in fact want you to review the code (but it doesn't need to be tonight!)

benmsanderson committed:
add changelog

benmsanderson committed:
revised changelog

maritsandstad commented on scripts/make_noresm_test.py:
I think this file might be ready for retirement tbh...

maritsandstad commented on scripts/make_pattern_4xco2_plots.py:
This one too?

maritsandstad commented on scripts/test_install.py:
I don't know how useful this file is anymore either...

maritsandstad commented on tests/unit/impacts/test_impacts_core.py:
Maybe this deserves a name change?

maritsandstad commented on tests/unit/impacts/test_ensemble.py:
```diff
@@ -0,0 +1,232 @@
+from meteor.impacts.impacts_core import ImpactCalculator, ImpactEnsemble, ImpactResult
+
+
+class MockCalculator(ImpactCalculator):
```
Do we need two implementations of this, or could we pull one from the other file?

benmsanderson replied:
pulling this into a common utils file

maritsandstad commented on tests/unit/test_cmip6_meteor_data_getter.py:
```diff
@@ -48,7 +96,329 @@ def test_get_unique_models():
+    assert "tas" in composite_data.data_vars
+
+
+def test_train_noise_model():
```
Strictly speaking this is more of an integration test than a unit test...

maritsandstad commented on tests/unit/test_cmip6_meteor_data_getter.py:
```diff
@@ -48,7 +96,329 @@ def test_get_unique_models():
+    assert noise_model.lag_order == 1
+
+
+def test_train_noise_model_with_cache():
```
Same comment as for previous test

maritsandstad replied:
Also, the two tests could be consolidated to save a bit of data getter time...

maritsandstad commented on tests/unit/test_cmip6_meteor_data_getter.py:
```diff
@@ -48,7 +96,329 @@ def test_get_unique_models():
+        assert noise_model2.fitted is True
+
+
+def test_train_noise_model_with_custom_temp():
```
Also something I would technically dub an integration test

maritsandstad commented on tests/unit/test_cmip6_meteor_data_getter.py:
```diff
@@ -48,7 +96,329 @@ def test_get_unique_models():
+        assert callable(train_noise_model_from_cmip6)
+
+
+def test_train_noise_model_picontrol_baseline():
```
Also integration test which should be consolidated...

maritsandstad commented on tests/unit/test_cmip6_meteor_data_getter.py:
```diff
@@ -48,7 +96,329 @@ def test_get_unique_models():
+    assert noise_model.fitted is True
+
+
+def test_error_handling():
```
Also ripe for consolidation with some of the other tests...

benmsanderson replied:
ok - done this in the integration move

maritsandstad commented on tests/unit/test_cmip6_meteor_data_getter.py:
```diff
@@ -48,7 +96,329 @@ def test_get_unique_models():
+@patch(
+    "meteor.cmip6_meteor_data_getter.Cmip6MeteorDataGetter.make_meteor_training_data_composite"
+)
+def test_make_meteor_training_data_composite_more_than_one(mock_composite):
```
Maybe I'm thick and it's late, but I don't really see what the mock_composite/fake_composite stuff is doing here...

benmsanderson replied:
agree - rewriting this test with real data

benmsanderson committed:
move some tests to integration, merged some tests, deleted outdated scripts

benmsanderson committed:
update version number

benmsanderson committed:
removed test-install

benmsanderson committed:
restored test_install
